examples/player/simid_player.js

```diff
@@ -609,14 +592,68 @@ class SimidPlayer {
     console.log('The creative has asked for the player to ping ' + requestedUrlArray);
   }
 
+  /**
+   * Called when video playback is complete.
+   * @private
+   */
+  videoComplete() {
+    this.simidProtocol.sendMessage(MediaMessage.ENDED);
+
+    if (this.requestedDuration_ == NO_REQUESTED_DURATION) {
+      this.stopAd(StopCode.MEDIA_PLAYBACK_COMPLETE);
+    }
+  }
+
+  /**
+   * Called when creative requests a change in duration of ad.
+   * @private
+   */
   onRequestChangeAdDuration(incomingMessage) {
-    if (this.requestedDuration_ != NO_REQUESTED_DURATION) {
-      // TODO: Support multiple change duration requests.
-      this.simidProtocol.reject(incomingMessage);
+    this.requestedDuration_  = incomingMessage.args['duration'];
+    //If requested duration is negative seconds, play entire ad
+    if (this.requestedDuration_ != UNLIMITED_DURATION && this.requestedDuration_< 0) {
+      this.sendLog("requested duration too short: " + this.requestedDuration_ + " seconds");
+      setTimeout(() => {
+        this.stopAd(StopCode.MEDIA_PLAYBACK_COMPLETE);
+        this.simidProtocol.reject(incomingMessage);
+      }, this.adVideoElement_.duration * 1000);
+    }//If requested duration is positive amount of seconds, call changeAdEndTime()
+    else {
+      this.changeAdEndTime();
+      this.simidProtocol.resolve(incomingMessage);
     }
-    const requestedDuration  = incomingMessage.args['duration'];
-    this.requestedDuration_ = requestedDuration;
-    this.simidProtocol.resolve(incomingMessage);
+  }
+
+  /**
+   * Compares the requested change duration with the current ad time
+   *  to determine when to stop the ad, every 250 milliseconds.
+   * @private
+   */
+  changeAdEndTime() {
+    this.interval_ = setInterval(() => {
+        this.compareAdAndRequestedDurations(); 
+    }, INTERVAL_TIMER);
+  }
+
+  /**
+   * Compares the duration of the ad with the requested change duration.
+   * If request duration is shorter, the ad stops early. 
+   * If the request duration is longer, the ad extends for the requested
+   *  amount of time.
+   * @private
+   */
+  compareAdAndRequestedDurations() {
+    if (this.adVideoElement_.currentTime >= this.requestedDuration_) {
```

Comment:
So what I was trying to explain earlier is that this logic is a little delicate.

For example, imagine this scenario:
currentTime is 10 seconds
videoTime is 20 seconds
RequestedTime is 30 seconds

The very first time we get into this section the statement this.adVideoElement_.currentTime >= this.requestedDuration_

will evaluate to false.

In javascript this means we will get into the else section
this.requestedDuration_ >= this.adVideoElement.duration

will evaluate to true.  This will then start the timer.

-----
Now imagine that the user pauses the ad with the pause button and makes a sandwich.  The ad will be terminated by the time they get back, which is not what we want. The desired effect is that unpausing should resume where they left off.

Instead consider this logic.

if (this.adVideoElement.ended) {
  // The video has ended but the creative has request more time, so set a timer to remove the creative for that case.
  clearInterval(this.interval_);
  if (this.requestedDuration_ == -2) {
   // The creative has asked for unlimited time, there will be no timeout. 
   // Note: We suggest any player allowing unlimited time to implement its own close ad button.  This will prevent bad user experiences.
  return;
  }
const durationChangeMs = (this.requestedDuration_ - this.adVideoElement_.duration) * 1000;
      setTimeout(() => {
        this.stopAd(StopCode.CREATIVE_INITIATED);
      }, durationChangeMs);
} else if (this.adVideoElement_.currentTime >= this.requestedDuration_) {
   // The creative has requested a shorted duration than the video's duration. This player will now stop the ad.
    clearInterval(this.interval_);
    this.stopAd(StopCode.CREATIVE_INITATED);
}

Reply:
understood